clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -534,15 +596,39 @@ public Subscription on(String target, Action callback) {
     }
 
     private class ConnectionState implements InvocationBinder {
-        HubConnection connection;
+        private HubConnection connection;
+        private AtomicInteger nextId = new AtomicInteger(0);
+        private HashMap<String, InvocationRequest> pendingInvocations = new HashMap<>();
 
         public ConnectionState(HubConnection connection) {
             this.connection = connection;
         }
 
+        public String getNextId() {
+            int i = nextId.incrementAndGet();
+            return Integer.toString(i);
+        }
+
+        public void cancelOutstandingInvocations(Exception ex) {
+            pendingInvocations.forEach((k, v) -> {
+                if (ex == null) {
+                    v.dispose();
+                } else {
+                    v.fail(ex);
+                }
+            });
+
+            pendingInvocations.clear();
```

Comment:
There's technically a race here right? If I call invoke in between the forEach and the clear, it will hang.

Reply:
Maybe we should just lock here.